File: (6 kyu) Reverse or rotate/(6 kyu) Reverse or rotate.ts
export class G964 {
  public static revrot(str: string, sz: number) {
    if (sz <= 0 || str == '' || sz > str.length) {
      return '';
    }
    const arr = [];
    const s = str.split('');
    while (s.length >= sz) {
      arr.push(s.splice(0, sz));
    }
    const res = arr.map((x) => {
      const sum = x.reduce((a: number, c: number) => a + Math.pow(c, 3), 0);
      if (sum % 2) {
        x.push(x[0]);
        x.shift();
        return x.join('');
      } else {
        return x.reverse().join('');
      }
    });
    return res.join('');
  }
}

File: 6kyu/(6 kyu) Reverse or rotate/(6 kyu) Reverse or rotate.cpp
#include <string>
#include <algorithm>
#include <numeric>
#include <math.h>

class RevRot
{
public:
    static std::string revRot(const std::string &strng, unsigned int sz)
    {
      if (sz == 0) { return ""; }
      std::string res = strng.substr(0, strng.size() / sz * sz);
  
      for (int i = 0; i < res.size(); i += sz)
      {
        auto begin = std::next(res.begin(), i);
        auto end   = std::next(res.begin(), i + sz);
        int sum = std::accumulate(begin, end, 0, RevRot::acc);
        if (sum % 2) { std::rotate(begin, begin + 1, end); }
        else { std::reverse(begin, end); }
      }
  
      return res;
    }
    
private:
    static int acc(int total, char &charNum)
    {
        return total + pow( static_cast<int>(charNum - 48), 3 );
    }
};
